### ISSUE TITLE
Memory pool weirdness

### ISSUE THREAD

**Issue by ad-tmpst** [tempesta-tech/tempesta#1540]
OK, so I've been reviewing the memory pool implementation and it does some really strange things that may be wrong.

I've started putting together the following test.
```c
	pl = __tfw_pool_new(1234);
	printk("pool_test: pool %px\n", pl);	// ffff888126ede010
	a = tfw_pool_alloc_not_align(pl, 1);
	b = tfw_pool_alloc(pl, 1);
	c = tfw_pool_alloc_not_align(pl, 1);
	printk("pool_test: a = %px\n", a);	// ffff888126ede020
	printk("pool_test: b = %px\n", b);	// ffff888126ede021
	printk("pool_test: c = %px\n", c);	// ffff888126ede029
	a = tfw_pool_alloc_not_align(pl, 1);
	printk("pool_test: a = %px\n", a);	// ffff888126ede02a
	b = tfw_pool_alloc_not_align(pl, 7);
	a = tfw_pool_realloc(pl, a, 1, 17);
	printk("pool_test: a = %px\n", a);	// ffff888126ede02a
```

So what I think happens here is:

`a` is 16 bytes past the pool pointer which makes sense since it's the first allocation. So far so good.

`b` goes right after `a`. Based on the function naming, I would expect the pointer to be aligned. The reason it works like this, is that in https://github.com/tempesta-tech/tempesta/blob/master/fw/pool.c#L125 if `align` is  true, then the size is adjusted to be a multiple of 8, but the returned pointer is not adjusted in any way, is this something expected? It surely isn't obvious.

The next allocation `c` isn't aligned either (although it shouldn't be), it just starts 8 bytes after the pointer to `b` (which we "aligned"). Still unsure what's being aligned here.

Finally, the `realloc` function assumes all sizes were previously aligned. https://github.com/tempesta-tech/tempesta/blob/master/fw/pool.c#L168 So in this case, it simply returns the previous pointer which we now can use to overwrite `b`.

I didn't go any further with this, there might be other problems.

I think it's possible to salvage `alloc` (assuming it actually should simply align pointers rather than... whatever it does.) but I'm not sure what to do with `realloc`. Overall my suggestion for fixing this is to remove the `align` flag entirely and always align pointers at a 8 byte boundary. I'm pretty sure this will almost never lead to a considerable overalignment in practice.

**Comment by krizhanovsky** [comment]
There is definitely an issue. The original implementation didn't use the `align` flag https://github.com/tempesta-tech/blog/blob/master/mempool/pool.c#L361 . Not aligned allocation was introduced with HTTP/2 hpack, which is memory hungry and performance critical.

The correct behavior must be to allocate aligned chunks by default. I see no reason why the size is aligned: the next allocation can always align the offset.

My understanding is that `__tfw_pool_realloc()` assumes that only the last allocation was aligned.

Actually, I'm wondering why don't we have a unit test for the pool? Please make a test, including your scenario, and make sure in another test that realloc correctly handles the last chunk, which can be reallocated just with `off` movement.

**Comment by ad-tmpst** [comment]
Fixed in #1541
